usr/src/uts/common/fs/zfs/spa_history.c

```diff
@@ -455,6 +455,7 @@ log_internal(nvlist_t *nvl, const char *operation, spa_t *spa,
 	fnvlist_add_string(nvl, ZPOOL_HIST_INT_NAME, operation);
 	fnvlist_add_uint64(nvl, ZPOOL_HIST_TXG, tx->tx_txg);
 
+	zfs_ereport_spa_history(spa, nvl);
```

Comment:
It provides a wealth of information that a userspace daemon can use to track the internal state of ZFS as an alternative to polling.  For example, at $WORK we use these for events to know when the following happen:
* zfs rename
* zpool create
* zpool import
* zpool set
* zfs create
* zfs destroy
* zfs snapshot

among many others.  We could implement separate events for all of them, but that would add much more code.  In FreeBSD, devd has no problem filtering the volume of events that this generates.  However, without this change a polling daemon gets unusably slow as the volume and snapshot count grows large.  But it's up to you.  Would you rather not add this feature to OpenZFS?